### PR TITLE
Nodejs: Improve Errors (5 of many)

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

This PR improves error/warn messages for nodejs.

## Out of Scope
- ParserError and CommandParserError are not covered (this is covered by https://github.com/fossas/fossa-cli/pull/801)

## Acceptance criteria

- When suboptimal strategy is used, warning with limitation is rendered. 

## Testing plan

- Analyze node project without lock file. 

## Risks

N/A

## References

Works on: https://github.com/fossas/team-analysis/issues/854

## Notes

I intend to have one refactor PR, once all error/warn diag are merged to have common actionable diagnostic format. 

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I updated `*schema.json` if I have made changes for `.fossa.yml`, `fossa-deps.{json, yaml, yml}`. You may also need to update these if you have added/removed new dependency (e.g. pip) or analysis target type (e.g. poetry).
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

**Review by csasarak — APPROVED** [review]
This looks okay to me, but I'm not as familiar with the needs/code base here. I'd recommend maybe having at least one other reviewer.

**Comment by meghfossa** [comment]
@scruffystuffs Tagging you as secondary

File: src/App/Docs.hs [hunk 1]
```diff
@@ -21,3 +22,6 @@ fossaYmlDocUrl = guidePathOf (maybe currentBranch ("v" <>) versionNumber) "/docs
 
 newIssueUrl :: Text
 newIssueUrl = sourceCodeUrl <> "/issues/new"
+
+strategyLangDocUrl :: Text -> Text
+strategyLangDocUrl path = guidePathOf (maybe currentBranch ("v" <>) versionNumber) ("/docs/references/strategies/languages/" <> path)
```
Comment (scruffystuffs) [inline src/App/Docs.hs:27]: `maybe currentBranch ("v" <>) versionNumber` should be a separate function in `App.Version`, I've seen it used quite a bit now.

File: src/Strategy/Node.hs [hunk 2]
```diff
@@ -138,6 +146,12 @@ analyzeNpmLock (Manifest file) graph = do
 
 analyzeNpm :: (Has Diagnostics sig m) => PkgJsonGraph -> m DependencyResults
 analyzeNpm wsGraph = do
+  _ <-
+    recover $
+      warnOnErr MissingEdges
+        . warnOnErr MissingDeepDeps
+        $ errCtx (MissingNodeLockFile) $ fatalText "Lock files - yarn.lock or package-lock.json were not discovered."
```
Comment (scruffystuffs) [inline src/Strategy/Node.hs:153]: [nit]: Use `void`, use `.` and `$` consistently, formatting.
```suggestion
  void
    . recover 
    . warnOnErr MissingEdges
    . warnOnErr MissingDeepDeps
    . errCtx (MissingNodeLockFile)
    $ fatalText "Lock files - yarn.lock or package-lock.json were not discovered."
```

File: src/Strategy/Node/Errors.hs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  yarnLockfileDocUrl,
+) where
+
+import App.Docs (strategyLangDocUrl)
+import Data.Text (Text)
+import Diag.Diagnostic (ToDiagnostic, renderDiagnostic)
+import Prettyprinter (Pretty (pretty), indent, vsep)
+
+yarnLockfileDocUrl :: Text
+yarnLockfileDocUrl = "https://classic.yarnpkg.com/lang/en/docs/yarn-lock/"
```
Comment (scruffystuffs) [inline src/Strategy/Node/Errors.hs:14]: We should include references to yarn v2 as well.  I don't think we'll have a way of knowing which one was used, so we should default to always showing both.

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
We may need a cycle on how to message both yarn v1 and v2, especially when we can't detect which one failed at this point.

Also, we need to add the errors for yarn and npm-lock as well.

**Comment by meghfossa** [comment]
@scruffystuffs Where we should add errors for yarn and npm-lock? Any errors due to parsing is already captured in other PR.

**Comment by scruffystuffs** [comment]
Looking again, it seems like the places that need still need error handling improvements are yarn v2 analysis and nodejs discovery.  Everything else is covered by parser error handling.

Discovery especially is a somewhat fragile process, which has to make a lot of assumptions about the way that the projects/workspaces fit together.  I don't know how much of that is relevant/useful for what we're doing here, but I think it needs to be addressed at some point.

**Comment by scruffystuffs** [comment]
Additionally, you may want to rework #811 to fit this model.

File: src/Strategy/Node/Errors.hs (hunk 3 above)
Comment (meghfossa) [reply src/Strategy/Node/Errors.hs:16]: fixed in: https://github.com/fossas/fossa-cli/pull/805/commits/c64f7b9247ca7f0d676674640ce83ccdda5c2e34

File: src/Strategy/Node.hs (hunk 2 above)
Comment (meghfossa) [reply src/Strategy/Node.hs]: fixed in https://github.com/fossas/fossa-cli/pull/805/commits/c64f7b9247ca7f0d676674640ce83ccdda5c2e34

File: src/App/Docs.hs (hunk 1 above)
Comment (meghfossa) [reply src/App/Docs.hs]: fixed in: https://github.com/fossas/fossa-cli/pull/805/commits/c64f7b9247ca7f0d676674640ce83ccdda5c2e34

File: src/Strategy/Node/Errors.hs [hunk 4]
```diff
@@ -13,12 +15,19 @@ import Prettyprinter (Pretty (pretty), indent, vsep)
...
+
 npmLockFileDocUrl :: Text
 npmLockFileDocUrl = "https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json"
 
 fossaNodeDocUrl :: Text
 fossaNodeDocUrl = strategyLangDocUrl "nodejs/nodejs.md"
 
+data CyclicPackageJson = CyclicPackageJson
+instance ToDiagnostic CyclicPackageJson where
+  renderDiagnostic (CyclicPackageJson) = "We detected cyclic references in package.json"
```
Comment (scruffystuffs) [inline src/Strategy/Node/Errors.hs:29]: Technically this is inaccurate as written:
How about: `We detected cyclic references between package.json files in the workspace.`

File: src/Strategy/Node.hs [hunk 5]
```diff
@@ -88,7 +96,7 @@ discover dir = context "NodeJS" $ do
       pure []
     else do
       globalGraph <- context "Building global workspace graph" $ pure $ buildManifestGraph manifestMap
-      graphs <- context "Splitting global graph into chunks" $ fromMaybeText "Detected cyclic package.json graph" $ splitGraph globalGraph
+      graphs <- context "Splitting global graph into chunks" $ fromMaybe CyclicPackageJson $ splitGraph globalGraph
```
Comment (scruffystuffs) [inline src/Strategy/Node.hs:99]: [todo]:  I don't like how I laid this error out, I think I just forgot to refactor it after making some change.
```suggestion
      -- TODO: refactor splitGraph to report which cycle we hit, not just report some unknown cycle
      graphs <- context "Splitting global graph into chunks" $ fromMaybe CyclicPackageJson $ splitGraph globalGraph
```

**Review by scruffystuffs — APPROVED** [review]
LGTM, one nit about error wording, and one todo for later

**Pull request merged**